---
title: "primerdesignR tutorial"
author: "Emily Humble"
date: "7/13/2016"
output: html_document
---

This `primerdesignR` package is developed to make it easy to design primers for a gene or region of interest. It will take a gene sequence, create and parse input files to programs used in a typical primer design workflow. It is designed to work for MacOSX and Linux. If you would like help using `primerdesignR`, please email emily.lhumble@gmail.com with any queries.

The minimum requirements are two input files:
1. A blast formatted genome 
2. A query sequence of interest 

It is also important that the following programs are installed and in the PATH.
`BLAST+ executables` 
`bedtools` 
`primer3` 

A [good tutorial](http://sfg.stanford.edu/BLAST.html) for installing and running BLAST
Bedtools [installation](http://bedtools.readthedocs.io/en/latest/content/installation.html)
Primer3 [manual](http://primer3.sourceforge.net/primer3_manual.htm#installMac) including installation instructions

The package can be installed using devtools:

```{r, eval=F}
# library(devtools)
# install_github("elhumble/primerdesignR")
```

### 1. Directory stucture.

This tutorial assumes that you are using `primerdesignR` for a project with a directory stucture similar to the following:

`seq_downloads`: query gene sequences from GenBank in closely related species
`primer_downloads`: published primers of gene of interest in closely related species
`blast_db`: formatted blast database
`blast_out`: directory for blast output
`bedtools`: directory for bedtools output
`primers`: directory for final primer3 output

We will design primers for sequencing the MC1R gene in the Antarctic fur seal genome. You will need to download the full draft fur seal genome from [Dryad](http://datadryad.org/resource/doi:10.5061/dryad.8kn8c) and format it into a BLAST database using the following command: `makeblastdb -in final.assembly.ArcGaz001_AP3.fasta -dbtype nucl -out blast_db/furseal`.

### 2. Retrieving query sequences

We will first download MC1R sequences from related species. You may also just have these saved as fasta files in the seq_downloads directory.

```{r}
library(primerdesignR)

# specify the NCBI accession numbers
# accessions <- c("CR541861.1", "CT010364.1", "AB028636.1", "NM_001004384.2", "XR_001521658.1", "XM_002199813.3")

# download the sequences

download_query_seqs <- function(accessions,
                     output){
        library(ape)
        seqs <- read.GenBank(accessions, species.names=T)
        names_sp <- data.frame(species = attr(seqs,"species"),
                         accs = names(seqs))
        names(seqs) <- attr(seqs, "species")
        write.dna(seqs, file = output, format="fasta", colsep = "")
        }

download_query_seqs(accessions = c("XM_546772", "AB725456", "NM_001114534", "JN575070"), 
         output = "data/mc1r_sp.fasta")

```

### 2. Blasting a query sequence against the genome.

This is done using the function `blastseq2genome` and requires the path to a directory containing the query sequence (or sequences) and the path to your genome database. The function will search for and blast all files ending in .fasta in the specified directory.

```{r}

blastseqs2genome(indir = "seq_downloads/", 
                 genomepath = "blast_db/furseal", 
                 outdir = "blast_out/")

```

In the directory `blast_out`, an output file `mc1r_seqs2genome` will have been created.

Sometimes, users may want to redesign a set of primers that already exist for a closely related species. To map primer sequences to a genome it is normally necessary to use less stringent alignment parameters. Let's see where a set of mc1r primers map to the fur seal genome.

```{r}

blastprimers2genome(indir = "~/bluetit_igf/seq_downloads/", 
                 genomepath = "~/blast_db/cyaCae2", 
                 outdir = "~/bluetit_igf/blast_out/",
                 )


```


